src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -715,8 +777,10 @@ impl Net {
         self.tx_rate_limiter.update_buckets(tx_bytes, tx_ops);
     }
 
-    fn read_tap(&mut self) -> std::io::Result<usize> {
-        self.tap.read(&mut self.rx_frame_buf)
+    pub(crate) fn read_tap(&mut self) -> std::io::Result<usize> {
+        let nr_iovecs = self.rx_buffer.parsed_descriptors[0].nr_iovecs as usize;
```

Comment:
We assume here that there are parsed descriptors. We do check that `read_from_mmds_or_tap` executes only with non empty `rx_buffers`. Maybe this should be `unsafe` with requirement of `rx_buffers` is not empty?
Or we can use `VecDeque::front` method to get optionally first element. And return some error if it is `None`.

Reply:
Yes, I agree. I'll go with the second approach and `unwrap` on in `read_from_mmds_or_tap` since at this point we do check that `rx_buffers` is not empty

Reply:
I went with the `unsafe` part of this, so that I avoid having to add yet another Error type.